modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryServerHandshakeManager.java

```diff
@@ -337,18 +323,20 @@ private void onHandshakeRejectedMessage(HandshakeRejectedMessage msg) {
         }
     }
 
-    private void sendHandshakeRejectedMessage(HandshakeRejectedMessage rejectionMessage, String reason) {
-        ChannelFuture sendFuture = channel.writeAndFlush(new OutNetworkObject(rejectionMessage, emptyList(), false));
+    private void rejectHandshakeDueToLosingClinch(RecoveryDescriptor descriptor) {
+        String localErrorMessage = "Failed to acquire recovery descriptor during handshake, it is held by: "
+                + descriptor.holderDescription();
 
-        NettyUtils.toCompletableFuture(sendFuture).whenComplete((unused, throwable) -> {
-            if (throwable != null) {
-                handshakeCompleteFuture.completeExceptionally(
-                        new HandshakeException("Failed to send handshake rejected message: " + throwable.getMessage(), throwable)
-                );
-            } else {
-                handshakeCompleteFuture.completeExceptionally(new HandshakeException(reason));
-            }
-        });
+        LOG.debug(localErrorMessage);
```

Comment:
Maybe LOG.isDebugEnabled() ?

Reply:
It is checked inside; also, handshakes happen seldom, we won't have any performance hit due to creating of an empty array each time.